userspace: compartmentalized app memory organization

jedomagalski opened the pull request:
# Summary

Initial attempt at addressing issue #6290.  The
following provides an alternative to using
CONFIG_APPLICATION_MEMORY by compartmentalizing data into
Memory Domains.  Dependent on MPU limitations, supports
compartmentalized Memory Domains for 1...N logical
applications.  This is considered an initial attempt at
designing flexible compartmentalized Memory Domains for
multiple logical applications.

# Overview

The current patch uses qualifiers to group data into
subsections.  The qualifier usage allows for dynamic subsection
creation and affords the developer a large amount of flexibility
in both the grouping, naming, and size of the resulting
partitions and domains that are built on these subsections. By
additional macro calls, functions are created that help calculate
the size, address, and permissions for the subsections and enable
the developer to control application data in specified partitions
and memory domains.

# Background

Initial attempts focused on creating a single
section in the linker script that then contained internally
grouped variables/data to allow MPU/MMU alignment and protection.
This did not provide additional functionality beyond
CONFIG_APPLICATION_MEMORY as we were unable to reliably group
data or determine their grouping via exported linker symbols.
Thus, the resulting decision was made to dynamically create
subsections using the current qualifier method. An attempt to
group the data by object file was tested, but found that this
broke applications such as ztest where two object files are
created: ztest and main.  This also creates an issue of grouping
the two object files together in the same memory domain while
also allowing for compartmenting other data among threads.

# Usage

 - Requires: app_memory/app_memdomain.h .
 - _app_dmem(id) marks a variable to be placed into a data
section for memory partition id.
 - _app_bmem(id) marks a variable to be placed into a bss
section for memory partition id.
 - These are seen in the linker.map as "data_smem_id" and
"data_smem_idb".
 - To create a k_mem_partition, call the macro
app_mem_partition(part0) where "part0" is the name then used to
refer to that partition. This macro only creates a function and
necessary data structures for the later "initialization".
 - To create a memory domain for the partition, the macro
app_mem_domain(dom0) is called where "dom0" is the name then
used for the memory domain.
 - To initialize the partition (effectively adding the partition
to a linked list), init_part_part0() is called. This is followed
by init_app_memory(), which walks all partitions in the linked
list and calculates the sizes for each partition.
 - Once the partition is initialized, the domain can be
initialized with init_domain_dom0(part0) which initializes the
domain with partition part0.
 - After the domain has been initialized, the current thread
can be added using add_thread_dom0(k_current_get()).

Example:

```
	/* create partition at top of file outside functions */
	app_mem_partition(part0);
	/* create domain */
	app_mem_domain(dom0);
	_app_dmem(dom0) int var1;
	_app_bmem(dom0) static volatile int var2;

	int main()
	{
		init_part_part0();
		init_app_memory();
		init_domain_dom0(part0);
		add_thread_dom0(k_current_get());
		...
	}
```

 - If multiple partitions are being created, a variadic
preprocessor macro can be used as provided in
app_macro_support.h:

	FOR_EACH(app_mem_partition, part0, part1, part2);

or, for multiple domains, similarly:

	FOR_EACH(app_mem_domain, dom0, dom1);

Similarly, the init_part_* can also be used in the macro:

	FOR_EACH(init_part, part0, part1, part2);

# Testing

 - This has been successfully tested on qemu_x86 and the
ARM frdm_k64f board.  See below for issues regarding
96b_carbon boards.
 - ARC changes are for compiling purposes only.  We don't
have hardware to facilitate testing on ARC.

# Current known issues

1) 96b_carbon board has one issue found in testing and one
possible other issue.
a) Specifically, we do not have sufficient regions to
support this patch's usage of memory domains because we
have maxed out MPU regions. Even if we disable
CONFIG_APPLICATION_MEMORY, its reserved region index does
not get reused. We believe that this is a separate issue
that can be fixed in a later patch.
b) The other issue that has not been tested but thought to
cause incompatibility with the 96b_carbon board is that
memory regions are not power of 2 aligned.  Specifically,
as the size of the subsection is unknown until runtime, the
ability to "buffer" the subsection without overlapping
another region to meet a power of 2 alignment is not
feesible (or at least unknown to the author).  Unless there
is another known method to force this either in a GNU
attribute or linker script, our current thought is this
will require a script to dynamically create the linker
section for this specific MPU.
2) checkpatch.pl lists several issues regarding the
following:
a) Complex macros. The FOR_EACH macros as defined in
app_macro_support.h are listed as complex macros needing
parentheses.  Adding parentheses breaks their
functionality, and we have otherwise been unable to
resolve the reported error.
b) __aligned() preferred. The _app_dmem_pad() and
_app_bmem_pad() macros give warnings that __aligned()
is preferred. Prior iterations had this implementation,
which resulted in errors due to "complex macros".
c) Trailing semicolon. The macro init_part(name) has
a trailing semicolon as the semicolon is needed for the
inlined macro call that is generated when this macro
expands.

Signed-off-by: Joshua Domagalski <jedomag@tycho.nsa.gov>

ruuddw reviewed (changes requested):
I didn't try on ARC yet, would expect it doesn't even compile because of the "Not implemented for this architecture" in app_memdomain.h for ARC.

In general I think it's good to simplify creation and allocation of variables to memory domain partitions, but the alignment and granularity restrictions for different MPU/MMUs have to be taken into account. Don't like the hardcoded ALIGNMENT_SIZE in app_memdomain, wouldn't this be defined already in MPU/MMU in a more generic way that could be used here?

Did you consider multiple, per thread, linker section instead of single app_smem for all? That might solve the alignment/granularity restrictions.

jedomagalski commented:
> I didn't try on ARC yet, would expect it doesn't even compile because of the "Not implemented for this architecture" in app_memdomain.h for ARC.
>
> In general I think it's good to simplify creation and allocation of variables to memory domain partitions, but the alignment and granularity restrictions for different MPU/MMUs have to be taken into account. Don't like the hardcoded ALIGNMENT_SIZE in app_memdomain, wouldn't this be defined already in MPU/MMU in a more generic way that could be used here?
> 
> Did you consider multiple, per thread, linker section instead of single app_smem for all? That might solve the alignment/granularity restrictions.

Regarding the ARC compilation, my mistake.  I had a previously built version of the patch that I didn't clear from the folder when I went to re-compile/check.  Also, we currently lack any actual ARC hardware.

For the hardcoded ALIGNMENT_SIZE, concur.  I had hardcoded them and left them to support testing and demonstrating this approach.  But, if were to be merged, would definitely want a more generic definition.

Regarding multiple, per-thread, linker-sections, somewhat unsure what you mean. If you are recommending static linker sections that are defined similarly to the manner that thread privilege stacks on ARM are defined, I don't see how that will help the alignment issue (aside from scripting).  Additionally, the goal of this patch was to allow for a granular sharing of globals among threads, rather than thread-specific storage or thread local storage...

andrewboie commented:
I haven't had time to review this yet but I will be soon, thanks for sending this.

jedomagalski commented:
Latest commit provides support for MPU architectures requiring power of 2 alignment.

SebastianBoe commented on include/app_memory/app_macro_support.h:
```diff
@@ -0,0 +1,32 @@
+#define FOR_EACH(x, ...) \
+	_GET_ARG(__VA_ARGS__, \
+	_for_10, _for_9, _for_8, _for_7, _for_6, _for_5, \
+	_for_4, _for_3, _for_2, _for_1, _for_0)(x, ##__VA_ARGS__)
```
This doesn't seem specific to app_memory, it would be unfortunate if a new component also required this and we had to either duplicate the macros or make references to app_memory.

Perhaps it would be better located closer to the other variadic preprocessor macros in zephyr/include/misc/util.h

jedomagalski committed:
userspace: compartmentalized app memory organization

Summary: Initial attempt at addressing issue 6290.  The
following provides an alternative to using
CONFIG_APPLICATION_MEMORY by compartmentalizing data into
Memory Domains.  Dependent on MPU limitations, supports
compartmentalized Memory Domains for 1...N logical
applications.  This is considered an initial attempt at
designing flexible compartmentalized Memory Domains for
multiple logical applications and, with the provided python
script and edited CMakeLists.txt, provides support for power
of 2 aligned MPU architectures.

Overview: The current patch uses qualifiers to group data into
subsections.  The qualifier usage allows for dynamic subsection
creation and affords the developer a large amount of flexibility
in the grouping, naming, and size of the resulting partitions and
domains that are built on these subsections. By additional macro
calls, functions are created that help calculate the size,
address, and permissions for the subsections and enable the
developer to control application data in specified partitions and
memory domains.

Background: Initial attempts focused on creating a single
section in the linker script that then contained internally
grouped variables/data to allow MPU/MMU alignment and protection.
This did not provide additional functionality beyond
CONFIG_APPLICATION_MEMORY as we were unable to reliably group
data or determine their grouping via exported linker symbols.
Thus, the resulting decision was made to dynamically create
subsections using the current qualifier method. An attempt to
group the data by object file was tested, but found that this
broke applications such as ztest where two object files are
created: ztest and main.  This also creates an issue of grouping
the two object files together in the same memory domain while
also allowing for compartmenting other data among threads.

Because it is not possible to know a) the name of the partition
and thus the symbol in the linker, b) the size of all the data
in the subsection, nor c) the overall number of partitions
created by the developer, it was not feasible to align the
subsections at compile time without using dynamically generated
linker script for MPU architectures requiring power of 2
alignment.

In order to provide support for MPU architectures that require a
power of 2 alignment, a python script is run at build prior to
when linker_priv_stacks.cmd is generated.  This script scans the
built object files for all possible partitions and the names given
to them. It then generates a linker file (app_smem.ld) that is
included in the main linker.ld file.  This app_smem.ld allows the
compiler and linker to then create each subsection and align to
the next power of 2.

Usage:
 - Requires: app_memory/app_memdomain.h .
 - _app_dmem(id) marks a variable to be placed into a data
section for memory partition id.
 - _app_bmem(id) marks a variable to be placed into a bss
section for memory partition id.
 - These are seen in the linker.map as "data_smem_id" and
"data_smem_idb".
 - To create a k_mem_partition, call the macro
app_mem_partition(part0) where "part0" is the name then used to
refer to that partition. This macro only creates a function and
necessary data structures for the later "initialization".
 - To create a memory domain for the partition, the macro
app_mem_domain(dom0) is called where "dom0" is the name then
used for the memory domain.
 - To initialize the partition (effectively adding the partition
to a linked list), init_part_part0() is called. This is followed
by init_app_memory(), which walks all partitions in the linked
list and calculates the sizes for each partition.
 - Once the partition is initialized, the domain can be
initialized with init_domain_dom0(part0) which initializes the
domain with partition part0.
 - After the domain has been initialized, the current thread
can be added using add_thread_dom0(k_current_get()).

  Example:
        /* create partition at top of file outside functions */
        app_mem_partition(part0);
        /* create domain */
        app_mem_domain(dom0);
        _app_dmem(dom0) int var1;
        _app_bmem(dom0) static volatile int var2;

        int main()
        {
                init_part_part0();
                init_app_memory();
                init_domain_dom0(part0);
                add_thread_dom0(k_current_get());
                ...
        }

 - If multiple partitions are being created, a variadic
preprocessor macro can be used as provided in
app_macro_support.h:

        FOR_EACH(app_mem_partition, part0, part1, part2);

or, for multiple domains, similarly:

        FOR_EACH(app_mem_domain, dom0, dom1);

Similarly, the init_part_* can also be used in the macro:

        FOR_EACH(init_part, part0, part1, part2);

Testing:
 - This has been successfully tested on qemu_x86 and the
ARM frdm_k64f board.  It compiles and builds power of 2
aligned subsections for the linker script on the 96b_carbon
boards.  These power of 2 alignments have been checked by
hand and are viewable in the zephyr.map file that is
produced during build. However, due to a shortage of
available MPU regions on the 96b_carbon board, we are unable
to test this.
 - When run on the 96b_carbon board, the test suite will
enter execution, but each individaul test will fail due to
an MPU FAULT.  This is expected as the required number of
MPU regions exceeds the number allowed due to the static
allocation. As the MPU driver does not detect this issue,
the fault occurs because the data being accessed has been
placed outside the active MPU region.
 - This now compiles successfully for the ARC boards
em_starterkit_em7d and em_starterkit_em7d_v22. However,
as we lack ARC hardware to run this build on, we are unable
to test this build.

Current known issues:
1) While the script and edited CMakeLists.txt creates the
ability to align to the next power of 2, this does not
address the shortage of available MPU regions on certain
devices (e.g. 96b_carbon).
2) checkpatch.pl lists several issues regarding the
following:
a) Complex macros. The FOR_EACH macros as defined in
app_macro_support.h are listed as complex macros needing
parentheses.  Adding parentheses breaks their
functionality, and we have otherwise been unable to
resolve the reported error.
b) __aligned() preferred. The _app_dmem_pad() and
_app_bmem_pad() macros give warnings that __aligned()
is preferred. Prior iterations had this implementation,
which resulted in errors due to "complex macros".
c) Trailing semicolon. The macro init_part(name) has
a trailing semicolon as the semicolon is needed for the
inlined macro call that is generated when this macro
expands.

Signed-off-by: Joshua Domagalski <jedomag@tycho.nsa.gov>

jedomagalski commented:
Latest commit moved the variadic macros to misc/util.h, enabled support to compile for ARC boards em_starterkit_em7d and em_starterkit_em7d_v22, and also added an align prior to each subsection in the linker script (for architectures where the MPU requires power of 2 sizing and size alignment) so that the start address is aligned.  Additionally, the "alignment size" now uses the _STACK_BASE_ALIGN for x86 and STACK_ALIGN for other architectures instead of hard-coded values at the top of app_memdomain.h .

stephensmalley commented:
Superseded by https://github.com/zephyrproject-rtos/zephyr/pull/7632
